week4 level1, level2

WonJoongLee opened the pull request:
Retrofit, Glide 라이브러리를 이용해 서버와 통신했습니다.

WonJoongLee committed:
(#10) 리드미에 week 4 정보 추가

- 첫 화면 리드미 표에 4주차 내용을 추가함.

WonJoongLee committed:
(#10) Retrofit, Gson 라이브러리 추가

WonJoongLee committed:
(#10) 인터넷 권한 추가

WonJoongLee committed:
(#10) Retrofit 위한 interface, data class 추가

WonJoongLee committed:
(#10) Retrofit Singleton pattern 위해 object 생성

WonJoongLee committed:
(#10) SignInActivity에서 Call, Callback 구현

WonJoongLee committed:
(#10) 폴더 정리 및 추가

WonJoongLee committed:
(#11) Github API Repository 수신 완료

- 내 레포지토리 받아오는 데까지 성공

WonJoongLee committed:
(#11) Toast Util 추가

- Toast message 띄울 일이 많아져 간편하게 사용하기 위해 추가했음

WonJoongLee committed:
(#11) 데이터 수신 로직 약간 변경

WonJoongLee committed:
(#11) 데이터 바인딩 이유 모를 오류 해결

- 리사이클러 뷰에서 스크롤을 해야 item 값들이 나오던 것을 해결했음. 해결 방법은 단순히 리사이클러뷰 아이템에서 데이터 바인딩 객체 <data>안에 있던 variable을 삭제해주는 것이 다였음.

WonJoongLee committed:
(#11) Progressbar 추가

- Repository 불러올 동안 Progress bar 표시

WonJoongLee committed:
(#11) github response data class 삭제

- 기존 Reposiotry data class로 받음.
- importantkeys.kt 삭제했음. 대신 xml string으로 관리함. gitignore에도 추가 완료.

WonJoongLee committed:
(#11) Design modification

- 약간의 디자인 추가 및 수정

WonJoongLee committed:
(#11) Github Follower API connected

WonJoongLee committed:
(#11) Add Glide for follower image

- 팔로워들 이미지 추가 위해 glide library 추가
- 팔로워 클릭하면 리플효과와 함께 해당 팔로워 레포지토리로 이동
- Recyclerview별로 item decoration 다르게 설정

l2hyunwoo commented on Fourth_Seminar/app/src/main/java/org/sopt/androidseminar/adapters/RepositoryAdapter.kt:
```diff
@@ -36,8 +40,32 @@ class RepositoryAdapter() :
             }
         }
+
+        private fun setSpanString(str: String): SpannableStringBuilder {
+            val noDescriptionStr = SpannableStringBuilder()
+            noDescriptionStr.append(str)
+            noDescriptionStr.setSpan(
+                StyleSpan(Typeface.ITALIC),
+                0,
+                noDescriptionStr.toString().length,
+                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
+            )
+            return noDescriptionStr
+        }
```
```suggestion
        private fun String.setSpan(): SpannableStringBuilder {
            val noDescriptionStr = SpannableStringBuilder()
            noDescriptionStr.append(this)
            noDescriptionStr.setSpan(
                StyleSpan(Typeface.ITALIC),
                0,
                this.length,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            return noDescriptionStr
        }
        
        "No Language".setSpan()
```

이런식으로 해도 뭔가 더 깔끔할 것 같기도 하고?

WonJoongLee replied:
그렇네 훨씬 난 듯

l2hyunwoo commented on Fourth_Seminar/.gitignore:
```diff
@@ -1,3 +1,5 @@
+app/src/main/res/values/secret_keys.xml
```
👍🏻

WonJoongLee replied:
처음써봄ㅎㅎ

l2hyunwoo commented on Fourth_Seminar/app/src/main/java/org/sopt/androidseminar/ui/FollowerFragment.kt:
```diff
@@ -21,32 +28,38 @@ class FollowerFragment : Fragment() {
+                response: Response<List<FollowerInfo>>
+            ) {
+                if (response.isSuccessful) {
+                    response.body()?.let {
+                        setFollowerRv(it)
+                    }
```
굳

l2hyunwoo reviewed (approved):
정말 고생 많이 했다. 서버통신 코드도 아주 좋아. 뭐라 더 말할 것도 없어

SeojinSeojin commented on Fourth_Seminar/app/src/main/AndroidManifest.xml:
```diff
@@ -2,13 +2,16 @@
         android:supportsRtl="true"
-        android:theme="@style/Theme.AndroidSeminar">
+        android:theme="@style/Theme.AndroidSeminar"
+        android:usesCleartextTraffic="true">
```
이건 뭐여?!?

l2hyunwoo replied:
@SeojinSeojin 너가 networkConfig 만들어준거하고 똑같은 기능을 하는거라고 보면 됨

SeojinSeojin commented on Fourth_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -42,7 +36,14 @@ class HomeActivity : AppCompatActivity() {
+                    //val noDescString : Spannable = "No Description".toSpannable()
+                    //noDescString.setSpan(StyleSpan(Typeface.ITALIC), 0, noDescString.toString().length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
+                    for(i in 0 until dataSize){
+                        repoList.add(RepositoryInfo(responsedData?.get(i)?.repoName?:"", responsedData?.get(i)?.repoDescription?:"No Description", responsedData?.get(i)?.repoLang?:""))
```
화려한 null처리가 코드를 감싸네~
repoName은 깃허브 API에서 필수로 넘겨주는 값이라 없으면 통신 에러가 있다는 것이니
통신 에러가 있어요~라고 알려주는 뭔가가 있으면 좋을 것 같아!!!

WonJoongLee replied:
맞아 이 부분 코드가 좀 더러워서 다시 수정했어!!

SeojinSeojin commented on Fourth_Seminar/app/src/main/java/org/sopt/androidseminar/ui/HomeActivity.kt:
```diff
@@ -56,11 +66,16 @@ class HomeActivity : AppCompatActivity() {
+            repoRecyclerView.adapter = repoAdapter
+            repoRecyclerView.setHasFixedSize(false)
+            Log.e("연결", "끝!")
+        }
```
데이터바인딩하는 스레드를 분리해준건가? 멋있구만
근데 runOnUiThread `{ setRepoRv() }`는 곧
runOnUiThread `{ runOnUiThread { --- } }` 이거와 같지 않아?
설명해준다면 고맙겠다!!

WonJoongLee replied:
이 때 오류 나서 이것저것 해본건데 결국 이 코드는 없어졌어 ㅋㅋ Thread 분리처럼 거창한거 아니고, 나는 UI Thread가 아닌  다른 Thread에서 Toast message를 띄워야 할 때만 써봤어ㅎ

SeojinSeojin commented on Fourth_Seminar/app/src/main/java/org/sopt/androidseminar/api/ServiceCreator.kt:
```diff
@@ -0,0 +1,23 @@
+            .build()
+    val soptService: SoptService = soptRetrofit.create(SoptService::class.java)
+    val repoService: RepoService = githubRetrofit.create(RepoService::class.java)
+}
```
오 이거 좋다!! 나도 이거 어떻게 같이 쓸지 고민했는데 잘 참고할게!

SeojinSeojin commented on Fourth_Seminar/app/src/main/java/org/sopt/androidseminar/util/RvItemDecoration.kt:
```diff
@@ -0,0 +1,20 @@
+            bottom = padding
+        }
+    }
+}
```
오 이렇게 하면 어떻게 생기게 될지 궁금하다..! 시간나면 캡처떠서 보여주세요~

WonJoongLee replied:
이렇게 리사이클러뷰 아이템 위아래로 간격이 들어가!!
![qwe](https://user-images.githubusercontent.com/57510192/118344681-848e2100-b56a-11eb-88fd-4d159fcc946c.PNG)

WonJoongLee committed:
(#11) Set color by parsing JSON

- 서진이가 준 JSON asset으로 레포지토리 언어 text에 색 추가함.
- 로그인 성공하면 Toast message 띄우는 부분에 "로그인 성공" 텍스트 추가

SeojinSeojin reviewed (commented):
잘짰다! 최고다!
리뷰보다는 질문을 더 많이했는데 답해준다면 감사감사!
수고많았어~

WonJoongLee committed:
(#10) Signup Call 객체 생성

WonJoongLee committed:
(#10) 회원가입 구현

WonJoongLee committed:
(#10) 마지막으로 리드미 추가

WonJoongLee committed:
(#10) 리드미 최종 수정

WonJoongLee committed:
(#10) 리드미 최종 수정
